modules/data-streams/src/main/java/org/elasticsearch/datastreams/action/TransportGetDataStreamsAction.java

```diff
@@ -146,11 +160,16 @@ static GetDataStreamAction.Response innerOperation(
         SystemIndices systemIndices,
         ClusterSettings clusterSettings,
         DataStreamGlobalRetentionSettings globalRetentionSettings,
+        DataStreamFailureStoreSettings dataStreamFailureStoreSettings,
         @Nullable Map<String, Long> maxTimestamps
     ) {
         List<DataStream> dataStreams = getDataStreams(state, indexNameExpressionResolver, request);
         List<GetDataStreamAction.Response.DataStreamInfo> dataStreamInfos = new ArrayList<>(dataStreams.size());
         for (DataStream dataStream : dataStreams) {
+            // For this action, we are returning whether the failure store is effectively enabled, either in metadata or by cluster setting.
+            // Users can use the get data stream options API to find out whether it is explicitly enabled in metadata.
+            boolean failureStoreEffectivelyEnabled = DataStream.isFailureStoreFeatureFlagEnabled()
+                && dataStream.isFailureStoreEffectivelyEnabled(dataStreamFailureStoreSettings);
```

Comment:
Yeah, I was a bit torn about the right place to do that check. I do have an `assert` in `DataStreamFailureStoreSetting.failureStoreEnabledForDataStreamName` to ensure that all calls are guarded by a feature flag check. I wondered whether we could move the check itself further up the stack.

The thing is that I don't think we can put this check into `DataStream.isFailureStoreExplicitlyEnabled`, because some of the callers don't want it, and I thought it would be confusing if `DataStream.isFailureStoreEffectivelyEnabled` did the check but `DataStream.isFailureStoreExplicitlyEnabled` didn't.

Also, I don't think we should sweat this decision too much, since wherever we put the check, we'll be taking it out before too long (🤞) when we remove the feature flag.

Let me know if you feel strongly that we should change this.